### ISSUE TITLE
Source style files the same way as .mdlrc files 

### ISSUE THREAD

**Issue by jamesquilty** [markdownlint/markdownlint#290]
## Description

At present it appears that`.mdlrc` files are sourced by walking up the directory tree from the current working directory, the directory where `mdl` was invoked, stopping at the user's home directory and uses the first `.mdlrc` file it finds on the way (#111). It appears that a style file  specified in `.mdlrc` is not sourced in the same manner, that `mdl` looks only in the current working directory and halts with a cryptic error message when it can't find it. This leads to `mdl` functioning in some directories and malfunctioning in others.

Edit: remove misleading and irrelevant filename.

### Environment

macOS 10.14 Mojave, which has ruby 2.3.7p456 as the system-installed version.

**MDL Version:** 0.5.0

### Expected Behaviour

`mdl` succeeds when the style file specified by filename only in a `.mdlrc` exists within in the directory hierarchy from the current working directory to the user's home directory, falling back to the default location(s). If the specified file can not be found `mdl` should either display a warning and continue or display a meaningful error and halt.

It will probably be beneficial to split relative file searches into two cases:

1. style file specified by filename only, i.e. specified style file does not contain `/` and does end in `.rb`, and
1. style file specified by relative path, i.e. specified style contains `/` and ends in `.rb`.

In case (1) it makes sense to walk the hierarchy starting from the current working directory, while in case (2) it makes more sense to just look relative to the location of the `.mdlrc` file.

### Actual Behaviour

`mdl` succeeds when the style file specified by relative path in `.mdlrc` exists relative to the current working directory but fails when invoked, for example, one level down

```shell
$ cd subdirectory
$ mdl *.md
~/.gem/ruby/2.3.0/gems/mdl-0.5.0/lib/mdl/style.rb:51:in `read': No such file or directory @ rb_sysopen - .mdlstyle.rb (Errno::ENOENT)
	from ~/.gem/ruby/2.3.0/gems/mdl-0.5.0/lib/mdl/style.rb:51:in `load'
	from ~/.gem/ruby/2.3.0/gems/mdl-0.5.0/lib/mdl.rb:26:in `run'
	from ~/.gem/ruby/2.3.0/gems/mdl-0.5.0/bin/mdl:10:in `<top (required)>'
	from ~/.gem/ruby/2.3.0/bin/mdl:22:in `load'
	from ~/.gem/ruby/2.3.0/bin/mdl:22:in `<main>'
```

## Replication Case

The following replicates the error:

```shell
mkdir -p mdlbugtest/subdirectory
cd mdlbugtest/
echo 'style ".mdlstyle.rb"' > .mdlrc
touch .mdlstyle.rb
touch subdirectory/somefile.md
mdl subdirectory/somefile.md
cd subdirectory/
mdl somefile.md
```

**Comment by jamesquilty** [comment]
An interesting related case is of a directory passed on the command-line which contains a `.mdlrc` file. At present `mdl directory` does not read `directory/.mdlrc`. I think that it probably should.

Edited to add: perhaps making Markdownlint attempt to read `.mdlrc` in each directory it visits as it's traversing a hierarchy (e.g. when performing a "git recurse") could be a way of allowing different rules for special directories, such as `.gitlab/issue_templates/` which I mentioned in #167.

**Comment by grantneufeld** [comment]
Yes. The current handling of the `style` setting in the `.mdlrc` file means we can only call `mdl` from the project directory with the `.mdlrc` file (typically the project root).

This makes integration with other tools suboptimal (I’m currently trying to integrate it with the text/code editors I use). We’re left with either running mdl against the whole project each time, or abandoning the use of styles (not really an option when the default rule configuration doesn’t meet the project’s needs). It also means that we can’t work with a file from a project in an editor that isn’t aware of the project (probably not a common use case, but I ran into that today).

**Comment by jaymzh** [comment]
in the short-term you know you can just pass in style files with `-s` and config files with `-c`, so that the directory you are in doesn't matter?

**Comment by jaymzh** [comment]
OK, so I wasn't super familiar with this part of the code. I went looking into this today.

The reason that sytles are loaded this way is because a "name" is *not* looking for a file at all. Per [the docs](https://github.com/markdownlint/markdownlint/blob/master/docs/configuration.md):

```
Note: The value for style_name must either end with .rb or have / in it in order to tell mdl to look for a custom style, and not a built-in style.
```

So, the idea is if you pass in a word, it's NOT looking for any file at all, it's looking for an *internal* style (we ship with several (`all`, `relaxed`,`cirosantilli`, `default`).

The thing I tend to do here is, in the relevant .mdlrc for the repo, do:

```ruby
style File.join(File.dirname(__FILE__), '.mdl.style')
```

In this way it doesn't not matter where you run it from, the style is always relative to the mdlrc that's found. This is quite useful.

The concern with walking up the path to find a `style` is that means the same style setting in two different stylerc's could mean something different and that seems, to me, far worse. It seems better to me, to be clear.

So, in terms of improvements we should probably make here, it seems to me:
1. We should have a far better error message, crashing in that way is clearly suboptimal.
2. We probably should add the above as a best-practice example on the config and styles pages.

**Comment by grantneufeld** [comment]
Oh—I hadn’t realized that the `.mdlrc` file is sourced as ruby code.
That means it’s a lot more flexible than I had understood.

Your proposed solution fixes the problem I was having. Thanks, much appreciated!

**Comment by jaymzh** [comment]
Yup, it's just a Mixlib::Config file ala Chef config or even Chef cookbooks `metadata.rb`

I'm glad it helped! Hopefully the attacked PR will mean others don't trip on this. Thanks for weighing in!

**Comment by jaymzh** [comment]
Somehow the PR I attached didn't include the doc fix I had made locally, so I fixed that. That should help as well.

**Comment by jaymzh** [comment]
Fixed in 0.9.0.
